[WIP] Use coroutineScope and async instead of CountDownLatch

mpeterss opened the pull request:
Using deferred to wait for all async calls to end instead of using
a CountDownLatch.

mpeterss committed:
Use Deferred instead of CountDownLatch

Using deferred to wait for all async calls to end instead of using
a CountDownLatch.

prasanthu reviewed (commented):
Try This instead 
[p1.txt](https://github.com/ostelco/ostelco-core/files/3846875/p1.txt)

prasanthu reviewed (approved):
Do as we discussed

mpeterss committed:
CoroutineScope insted of deferred

The CoroutineScope will wait for all Jobs to finish so there is
no need for the Deferred.

prasanthu committed:
Merge branch 'develop' into feature/ocs-coroutines

mpeterss commented:
Do not merge this, I want to test this more to be sure it is working correctly

vihangpatil reviewed (changes requested):
DO NOT MERGE THIS.  The Kotlin Coroutine might not work.

The scope will work when the child coroutines are async using suspend.
In that case, the callback is handled as a continuation.

But in this case, the callback is an execution of the lambda function which is passed as the last parameter.

So, the parent coroutine scope will be blocked till all the coroutines in it are completed, but they get completed immediately.  The scope will not be blocked until all the callbacks are executed.

prasanthu commented:
> DO NOT MERGE THIS. The Kotlin Coroutine might not work.
> 
> The scope will work when the child coroutines are async using suspend.
> In that case, the callback is handled as a continuation.
> 
> But in this case, the callback is an execution of the lambda function which is passed as the last parameter.
> 
> So, the parent coroutine scope will be blocked till all the coroutines in it are completed, but they get completed immediately. The scope will not be blocked until all the callbacks are executed.

Are you sure ?, @mpeterss Checked this using AT.

prasanthu commented:
> DO NOT MERGE THIS. The Kotlin Coroutine might not work.
> 
> The scope will work when the child coroutines are async using suspend.
> In that case, the callback is handled as a continuation.
> 
> But in this case, the callback is an execution of the lambda function which is passed as the last parameter.
> 
> So, the parent coroutine scope will be blocked till all the coroutines in it are completed, but they get completed immediately. The scope will not be blocked until all the callbacks are executed.

Are you talking about the bimap lamda? I don't see any asynchronous calls than the `async` in  `consumeRequestHandler` . If the whole loop (`request.msccList.forEach`) runs synchronously, the asyncs methods will be called before the loop is finished. In such cases the async calls are part of the `coroutineScope`

vihangpatil commented on ocs-ktc/src/main/kotlin/org/ostelco/prime/ocs/core/OnlineCharging.kt:
```diff
@@ -96,57 +94,56 @@ object OnlineCharging : OcsAsyncRequestConsumer {
+                        storage.consume(
+                                msisdn = consumptionRequest.msisdn,
+                                usedBytes = consumptionRequest.usedBytes,
+                                requestedBytes = consumptionRequest.requestedBytes) { storeResult ->
```
I don't think that the coroutine scope will wait til the last argument to `storage.consume` - the lambda with storeResult arg is called back.

mpeterss replied:
correct. I have seen it fail in AT. So this is not working

prasanthu replied:
Now I understand what you are talking about

prasanthu replied:
`storage.consume()` calls it callback in a different `CoroutineScope`. A new one is created when we call `writeSuspended()`

vihangpatil committed:
Updated Neo4j Load Test

vihangpatil committed:
Removed CountDownLatch from Neo4j Load Test

mpeterss committed:
Fix check user coroutine

This was not waiting for the callback to complete. Now it use
async and scope.

prasanthu committed:
Convert CompletionStage to Deferred

This will help to avoid confusion of while using threads and
coroutines together.

prasanthu committed:
Convert CompletionStage to Deferred

This will help to avoid confusion of while using threads and
coroutines together.

prasanthu committed:
Merge branch 'develop' into feature/ocs-coroutines

prasanthu reviewed (commented):
Review my changes

prasanthu committed:
Revert "Convert CompletionStage to Deferred"

This reverts commit 725192bb6f8d33049ac930640c724560b8405dfb.
The actual patch fails to run `load test Neo4j`

prasanthu commented:
I have reverted the change I made for converting CompletionStage to Deferred. The load test with `10_000` requests failed with this change. The max I could do was `500`

prasanthu commented:
That revert made the `creditControlRequestInitTerminateNoCredit` fail. So bringing the change back. We need to sit together and understand this

vihangpatil committed:
Merge branch 'develop' into feature/ocs-coroutines

vihangpatil committed:
Merge branch 'develop' into feature/ocs-coroutines
